dotnet/src/Connectors/Connectors.Memory.Qdrant/QdrantVectorStoreRecordCollection.cs

```diff
@@ -235,103 +236,120 @@ public virtual Task DeleteCollectionAsync(CancellationToken cancellationToken =
+                Payload = { }
+            };
+
+            foreach (KeyValuePair<string, Value> payloadEntry in retrievedPoint.Payload)
+            {
+                pointStruct.Payload.Add(payloadEntry.Key, payloadEntry.Value);
+            }
+
+            yield return VectorStoreErrorHandler.RunModelConversion(
+                DatabaseName,
                 this._collectionName,
-                key,
-                wait: true,
-                cancellationToken: cancellationToken));
+                OperationName,
+                () => this._mapper.MapFromStorageToDataModel(pointStruct, new() { IncludeVectors = includeVectors }));
+        }
     }
 
     /// <inheritdoc />
-    public virtual Task DeleteAsync(Guid key, CancellationToken cancellationToken = default)
+    public virtual Task DeleteAsync(TKey key, CancellationToken cancellationToken = default)
     {
         Verify.NotNull(key);
 
         return this.RunOperationAsync(
             DeleteName,
-            () => this._qdrantClient.DeleteAsync(
-                this._collectionName,
-                key,
-                wait: true,
-                cancellationToken: cancellationToken));
+            () => key switch
+            {
+                ulong id => this._qdrantClient.DeleteAsync(this._collectionName, id, wait: true, cancellationToken: cancellationToken),
+                Guid id => this._qdrantClient.DeleteAsync(this._collectionName, id, wait: true, cancellationToken: cancellationToken),
+                _ => throw new NotSupportedException($"The provided key type '{key.GetType().Name}' is not supported by Qdrant.")
+            });
     }
 
     /// <inheritdoc />
-    public virtual Task DeleteAsync(IEnumerable<ulong> keys, CancellationToken cancellationToken = default)
+    public virtual Task DeleteAsync(IEnumerable<TKey> keys, CancellationToken cancellationToken = default)
     {
         Verify.NotNull(keys);
 
-        return this.RunOperationAsync(
-            DeleteName,
-            () => this._qdrantClient.DeleteAsync(
-                this._collectionName,
-                keys.ToList(),
-                wait: true,
-                cancellationToken: cancellationToken));
-    }
+        IList? keyList = null;
 
-    /// <inheritdoc />
-    public virtual Task DeleteAsync(IEnumerable<Guid> keys, CancellationToken cancellationToken = default)
-    {
-        Verify.NotNull(keys);
+        switch (keys)
+        {
+            case IEnumerable<ulong> k:
+                keyList = k.ToList();
+                break;
+
+            case IEnumerable<Guid> k:
+                keyList = k.ToList();
+                break;
+
+            case IEnumerable<object> k:
+                var objectKeys = k.ToList();
+                keyList = objectKeys[0] switch
+                {
+                    ulong id => objectKeys.Cast<ulong>().ToList(),
+                    Guid id => objectKeys.Cast<Guid>().ToList(),
```

Comment:
The `ToList` is being called twice for `IEnumerable<object>`, moreover we risk `IndexOutOfRangeException` by accessing `[0]`.

Reply:
You're right, I was sloppy here. Will reimplement better.